src/GraphQL/Internal/Validation.hs

```diff
@@ -156,17 +156,17 @@ validate schema (AST.QueryDocument defns) = runValidator $ do
     splitDefns (AST.DefinitionFragment frag) = Right frag
 
     splitOps (AST.AnonymousQuery ss) = Left ss
-    splitOps (AST.Query node@(AST.Node name _ _ _)) = Right (name, (Query, node))
-    splitOps (AST.Mutation node@(AST.Node name _ _ _)) = Right (name, (Mutation, node))
+    splitOps (AST.Query node@(AST.Node maybeName _ _ _)) = Right (maybeName, (Query, node))
+    splitOps (AST.Mutation node@(AST.Node maybeName _ _ _)) = Right (maybeName, (Mutation, node))
 
-    assertAllFragmentsUsed :: Fragments value -> Set Name -> Validation ()
+    assertAllFragmentsUsed :: Fragments value -> Set (Maybe Name) -> Validation ()
     assertAllFragmentsUsed fragments used =
-      let unused = Map.keysSet fragments `Set.difference` used
+      let unused = ( Set.map pure (Map.keysSet fragments)) `Set.difference` used
```

Comment:
Are names of fragments optional?

Reply:
They are not. What's happening here is that we've changed the type of the "observed names" set. I don't think it's strictly necessary (you can't have anonymous fragments, so there's no point recording that we've seen `Nothing`), but it only introduces a minor error message regression that we can fix easily enough later, leaving this PR a net improvement.

Reply:
There were two possibilities here, if I understood things correctly.
1. Leave Node Names packed in a Maybe, with the consequence that the accumulated state of StateT must change to contain "Maybe Name" as well. Since names into StateT can also come from Fragments, those names also needed to be packed into Maybes (this seemed like the safest approach to me, and although it does change error messages associated with fragments to imply that Fragment Names can be Nothing [Is this the "minor error message regression"?], in actuality, they never would be). 
2. Unpack Node Names...
    a. Under the assumption that Nodes MUST be named in this context (I was not at all sure of this) -> throw an error if Nothing.
    b. Under the assumption that Node Names can be Nothing in this context -> No way to proceed (or rather, only hacky ways to proceed, i.e. Set name to "").

Reply:
That's mostly right.

The `StateT` is there to track which fragments have been visited. Since all fragments have names, it's impossible for a node without a name visit a fragment. That means if we find an anonymous node, we don't need to update the `StateT`.

(Multi-tasking heavily at the moment, so I might be horribly wrong).